File: sdks/local-rp/ruby/test/test_conformance_callback_box.rb
# frozen_string_literal: true

require_relative 'test_helper'

class TestConformanceCallbackBox < Minitest::Test
  include ConformanceHelper

  def open_case(c)
    encrypted = LinkkeysLocalRp::Types::LocalRpEncryptedCallback.new(
      header: ConformanceHelper.hex(c['header_cbor_hex']),
      ciphertext: ConformanceHelper.hex(c['ciphertext_hex'])
    )
    allowed_suites = c['allowed_suites']
    decrypt_key = ConformanceHelper.hex(c['decrypt_private_key_hex'])
    LinkkeysLocalRp::LocalRp.open_local_rp_callback(encrypted, decrypt_key, allowed_suites)
  end

  def test_positive_cases
    callback_box_vector['positive_cases'].each do |c|
      _header, signed_payload = open_case(c)
      assert_equal c['plaintext_cbor_hex'], signed_payload.to_cbor.unpack1('H*'), "suite #{c['suite']}"
    end
  end

  def test_negative_cases
    callback_box_vector['negative_cases'].each do |c|
      assert_raises(StandardError, c['name']) { open_case(c) }
    end
  end

  def test_kdf_context_matches
    callback_box_vector['positive_cases'].each do |c|
      suite = c['suite']
      ephemeral_public = ConformanceHelper.hex(c['ephemeral_public_key_hex'])
      recipient_public = ConformanceHelper.hex(c['recipient_public_key_hex'])
      tag = 'linkkeys-local-rp-callback-box'.b
      expected_context = tag + suite.b + ephemeral_public + recipient_public
      assert_equal c['kdf_context_hex'], expected_context.unpack1('H*')
      assert_equal c['aad_hex'], (expected_context + ConformanceHelper.hex(c['header_cbor_hex'])).unpack1('H*')
    end
  end
end

File: sdks/local-rp/python/tests/test_conformance_callback_box.py
"""callback_box.json: the local-RP callback sealed box, per suite, positive
and negative (13) cases -- header-field-flip AAD tamper, unadvertised suite,
unknown suite id, low-order ephemeral key, wrong recipient key, truncated
ciphertext."""

import pytest
from conftest import hex_decode

from linkkeys_local_rp import crypto
from linkkeys_local_rp.generated.types import LocalRpEncryptedCallback
from linkkeys_local_rp.local_rp import open_local_rp_callback


def _allowed_suites(case: dict):
    return [crypto.AeadSuite.parse_str(s) for s in case["allowed_suites"]]


def test_callback_box_positive_cases_open(callback_box):
    cases = callback_box["positive_cases"]
    assert len(cases) == 2

    for case in cases:
        header_bytes = hex_decode(case["header_cbor_hex"])
        ciphertext = hex_decode(case["ciphertext_hex"])
        decrypt_key = hex_decode(case["decrypt_private_key_hex"])
        allowed = _allowed_suites(case)

        encrypted = LocalRpEncryptedCallback(header=header_bytes, ciphertext=ciphertext)
        header, signed_payload = open_local_rp_callback(encrypted, decrypt_key, allowed)

        assert header.suite == case["suite"]
        assert header.fingerprint == case["fingerprint"]
        assert header.nonce == hex_decode(case["nonce_hex"])
        assert header.state == hex_decode(case["state_hex"])
        assert header.issued_at == case["issued_at"]
        assert header.expires_at == case["expires_at"]

        plaintext = signed_payload.to_cbor()
        assert plaintext == hex_decode(case["plaintext_cbor_hex"])


def test_callback_box_negative_cases_fail(callback_box):
    cases = callback_box["negative_cases"]
    assert len(cases) == 13

    for case in cases:
        header_bytes = hex_decode(case["header_cbor_hex"])
        ciphertext = hex_decode(case["ciphertext_hex"])
        decrypt_key = hex_decode(case["decrypt_private_key_hex"])
        allowed = _allowed_suites(case)

        encrypted = LocalRpEncryptedCallback(header=header_bytes, ciphertext=ciphertext)
        with pytest.raises(Exception):
            open_local_rp_callback(encrypted, decrypt_key, allowed)
